### PR TITLE
helm chart inflator function in go

### PR DESCRIPTION
**Pull request by natasha41575**
examples and more tests (especially for values file and local charts) are coming in a followup PR

most of the implementation is taken from https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go, and tweaked so that it can be run in a container environment.

### CODE REVIEW

**Comment by natasha41575** [comment]
@mengqiy Does it make sense to add more examples for kustomize? If so, where should they go?

File: functions/go/inflate-helm-chart/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  repo: string
+  releaseName: string
+  namespace: string
+  valuesInline: map[string]interface{}
+  valuesFile: string
+  valuesMerge: string
+  includeCRDs: bool
+```
+
+#### Parameter Descriptions
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:66]: I'd suggest you to merge this section with the "schema" above, since the 2 formats of functionConfig are not the same. The CRD one has more fields and the fields can be more than string.

File: functions/go/inflate-helm-chart/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+`version`      | The version of the chart
+`repo`         | A URL locating the chart on the internet
+`releaseName`  | Replaces RELEASE_NAME in the chart template output
+`namespace`    | Sets the target namespace for a release (.Release.Namespace in the template)
+`valuesInline` | Values to use instead of default values that accompany the chart
+`valuesFile`   | ValuesFile is local file path to a values file to use _instead of_ the default values that accompanied the chart. The default values are in '{ChartHome}/{Name}/values.yaml'.
+`valuesMerge`  | ValuesMerge specifies how to treat ValuesInline with respect to Values. Legal values: 'merge', 'override' (default), 'replace'.
+`includeCRDs`  | Specifies if Helm should also generate CustomResourceDefinitions, defaults to 'false'
+
+#### Example with kustomize
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:84]: Please follow our usage doc template (https://raw.githubusercontent.com/GoogleContainerTools/kpt-functions-catalog/master/functions/go/_template/README.md):

```
## Examples

<!--mdtogo:Examples-->
Your content here...
<!--mdtogo-->
```

There are 2 options (they are not mutually exclusive) for the example(s) using kustomize:
- Add example(s) in the `examples` directory. We can potentially test it in the CI.
- Add inline example(s) in the usage doc.

If we do the later, I'd suggest to add an small inline example for kpt as well.
@natasha41575 WDYT?

File: functions/go/inflate-helm-chart/helm_chart_inflator_generator.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
+// Copyright 2021 The Kubernetes Authors.
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/helm_chart_inflator_generator.go:1]: Are these code copied from kustomize?
If so, we should keep in `third_party` like https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/functions/go/starlark/third_party and https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/functions/ts/kubeval/third_party

**Review by mengqiy — COMMENTED** [review]
Please rebase on the latest HEAD. We now expected the folder named `generated` to be checked in.

**Comment by mengqiy** [comment]
Re. CI failure, please take a look at https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/CONTRIBUTING.md#contribute-new-functions

File: functions/go/inflate-helm-chart/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  && tar -xf helm-v3.6.3-linux-amd64.tar.gz \
+  && mv ./linux-amd64/helm /usr/local/bin/helmV3 \
+  && chmod +x /usr/local/bin/helmV3
+
+#############################################
+
+FROM alpine:3.13
+COPY --from=0 /usr/local/bin/function /usr/local/bin/function
+COPY --from=0 /usr/local/bin/helmV3 /usr/local/bin/helmV3
+
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/Dockerfile:27]: I don't think it's necessary but it is a good idea to do so, so I added it

File: functions/go/inflate-helm-chart/README.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+### FunctionConfig
+
+<!--mdtogo:Long-->
+
+There are 2 kinds of `functionConfig` supported by this function:
+
+- `ConfigMap`
+- A custom resource of kind `HelmChartInflator`
+
+To use a `ConfigMap` as the `functionConfig`, the desired parameters must be
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:56]: I have updated the kustomize examples to have one example for each functionConfig format, and added a sentence to each format explaining that only `name` is required

File: examples/inflate-helm-chart-local/.expected/exec.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+kpt fn eval --image-pull-policy ifNotPresent --image gcr.io/kpt-fn/inflate-helm-chart:unstable \
+--mount type=bind,src="$(pwd)",dst=/tmp/charts -- \
+name=helloworld-chart \
+releaseName=test \
+valuesFile=tmp/charts/helloworld-values/values.yaml
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply examples/inflate-helm-chart-local/.expected/exec.sh]: done

File: examples/inflate-helm-chart-local/.expected/diff.patch [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
++apiVersion: v1
++kind: ServiceAccount
++metadata:
++  name: test-helloworld-chart
++  labels:
++    helm.sh/chart: helloworld-chart-0.1.0
++    app.kubernetes.io/name: helloworld-chart
++    app.kubernetes.io/instance: test
++    app.kubernetes.io/version: "1.16.0"
++    app.kubernetes.io/managed-by: Helm
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply examples/inflate-helm-chart-local/.expected/diff.patch]: added a newline

File: examples/inflate-helm-chart-local/helloworld-chart/.helmignore [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# Patterns to ignore when building packages.
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply examples/inflate-helm-chart-local/helloworld-chart/.helmignore]: removed the file

File: functions/go/inflate-helm-chart/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,23 @@
+FROM golang:1.16-alpine3.12
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/Dockerfile]: changed to use golang:1.16-alpine3.13

File: functions/go/inflate-helm-chart/helm_chart_inflator_generator.go [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+func (p *plugin) absChartHome() string {
+	path, _ := filepath.Abs(p.ChartHome)
+	return path
+}
+
+func (p *plugin) runHelmCommand(
+	args []string) ([]byte, error) {
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	cmd := exec.Command("helmV3", args...)
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/helm_chart_inflator_generator.go]: changed to `helm`

File: functions/go/inflate-helm-chart/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+WORKDIR /go/src/
+
+COPY go.mod go.sum ./
+RUN go mod download
+
+COPY . .
+RUN go build -o /usr/local/bin/function ./
+
+RUN apk update && apk add --no-cache wget
+RUN wget https://get.helm.sh/helm-v3.6.3-linux-amd64.tar.gz \
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/Dockerfile]: done

File: functions/go/inflate-helm-chart/README.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+<!--mdtogo-->
+
+You can learn more about helm charts [here][helm].
+
+## Usage
+
+This function can be used with any KRM function orchestrators (e.g. kpt or kustomize) 
+to inflate a specified Helm chart.
+It can be used imperatively or declaratively when run with kustomize. With kpt, it can be
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: done

File: functions/go/inflate-helm-chart/README.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+```yaml
+data:
+  chartHome: string
+  configHome: string
+  name: string
+  version: string
+  repo: string
+  releaseName: string
+  namespace: string
+  valuesInline: map[string]interface{}
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: updated it to match the design doc

File: functions/go/inflate-helm-chart/README.md [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  valuesFile: string
+  valuesMerge: string
+  includeCRDs: bool
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: changed it

File: functions/go/inflate-helm-chart/README.md (hunk 1 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: I merged it to clarify the differences between the two functionConfigs

File: functions/go/inflate-helm-chart/README.md (hunk 2 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: I added inline examples, two for kustomize and one for kpt. The other kpt examples are still under `/examples`, LMK what you think

File: functions/go/inflate-helm-chart/helm_chart_inflator_generator.go (hunk 3 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/helm_chart_inflator_generator.go]: I moved the copied code to `third_party`

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go [hunk 15]
```diff
@@ -0,0 +1,336 @@
+// Copyright 2021 The Kubernetes Authors.
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go:1]: Since we fork the code, I guess there are some modification, right?
When we fork code, we usually do it in a separate PR. If possible, it should simply copy the code w/o any modification.
Any modification should be done in a separate PR.

File: functions/go/inflate-helm-chart/go.mod [hunk 16]
```diff
@@ -3,8 +3,10 @@ module github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/inflat
 go 1.16
 
 require (
-	github.com/imdario/mergo v0.3.12
+	github.com/imdario/mergo v0.3.5
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/go.mod:6]: Why downgrade it?

File: functions/go/inflate-helm-chart/go.mod [hunk 17]
```diff
@@ -3,8 +3,10 @@ module github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/inflat
...
 	github.com/pkg/errors v0.9.1
-	sigs.k8s.io/kustomize/api v0.8.11
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
+	sigs.k8s.io/kustomize/api v0.8.11-0.20210614195535-7e8ba62e9fd9
+	sigs.k8s.io/kustomize/kyaml v0.10.21
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/go.mod:10]: Why not use the latest?

File: functions/go/inflate-helm-chart/go.sum [hunk 18]
```diff
@@ -215,19 +218,21 @@ gopkg.in/yaml.v2 v2.2.1/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
...
 k8s.io/klog/v2 v2.0.0/go.mod h1:PBfzABfn139FHAV07az/IF9Wp1bkk3vpT2XSJ76fSDE=
 k8s.io/kube-openapi v0.0.0-20210421082810-95288971da7e h1:KLHHjkdQFomZy8+06csTWZ0m1343QqxZhR2LJ1OxCYM=
 k8s.io/kube-openapi v0.0.0-20210421082810-95288971da7e/go.mod h1:vHXdDvt9+2spS2Rx9ql3I8tycm3H9FDfdUoIuKCefvw=
-sigs.k8s.io/kustomize/api v0.8.11 h1:LzQzlq6Z023b+mBtc6v72N2mSHYmN8x7ssgbf/hv0H8=
-sigs.k8s.io/kustomize/api v0.8.11/go.mod h1:a77Ls36JdfCWojpUqR6m60pdGY1AYFix4AH83nJtY1g=
-sigs.k8s.io/kustomize/kyaml v0.11.0 h1:9KhiCPKaVyuPcgOLJXkvytOvjMJLoxpjodiycb4gHsA=
-sigs.k8s.io/kustomize/kyaml v0.11.0/go.mod h1:GNMwjim4Ypgp/MueD3zXHLRJEjz7RvtPae0AwlvEMFM=
+sigs.k8s.io/kustomize/api v0.8.11-0.20210614195535-7e8ba62e9fd9 h1:ge8BT4wAOMQRJ6wL5E7/d4vyI+K3ckJf40EbEyggvWA=
+sigs.k8s.io/kustomize/api v0.8.11-0.20210614195535-7e8ba62e9fd9/go.mod h1:C431Rq3Ne1LxTCWA93ZN1R5v0KxioCGVlLVaUlFDvJw=
+sigs.k8s.io/kustomize/kyaml v0.10.20/go.mod h1:TYWhGwW9vjoRh3rWqBwB/ZOXyEGRVWe7Ggc3+KZIO+c=
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/go.sum:233]: Please ensure you have run `go mod tidy` to trim unnecessary items.

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go [hunk 19]
```diff
@@ -196,16 +174,32 @@ func (p *HelmChartInflationGeneratorPlugin) replaceValuesInline() error {
 }
 
 // copyValuesFile to avoid branching.  TODO: get rid of this.
-func (p *HelmChartInflationGeneratorPlugin) copyValuesFile() (string, error) {
-	b, err := p.h.Loader().Load(p.ValuesFile)
+func (p *Plugin) copyValuesFile() (string, error) {
+	var b []byte
+	var err error
+	b, err = ioutil.ReadFile(p.ValuesFile)
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go:180]: If it's a local file, does it require file scheme (e.g. file:///tmp/my-values.yaml)?

File: functions/go/inflate-helm-chart/README.md [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+  name: string
+  version: string
+  repo: string
+  releaseName: string
+  namespace: string
+  valuesFile: string
+  includeCRDs: string
+```
+
+| Field        |  Description |
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:72]: For each field, we should have example values if possible.
E.g. for `valuesFile` should it always be a URL?

File: functions/go/inflate-helm-chart/README.md [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+    repo: https://itzg.github.io/minecraft-server-charts
+    version: 3.1.3
+    releaseName: test
+```
+
+For both of the above kustomizations, the helm chart can be
+inflated with the following command:
+
+```shell
+kustomize build --enable-alpha-plugins --network .
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:199]: If a minimum kustomize is required, please specify here.

File: functions/go/inflate-helm-chart/main.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
+package main
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/main.go:1]: Please add Google license header

File: functions/go/inflate-helm-chart/main.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
+	kyaml "sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+//nolint
+func main() {
+	asp := HelmChartProcessor{}
+	cmd := command.Build(&asp, command.StandaloneEnabled, false)
+
+	cmd.Short = "inflate helm chart"
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/main.go:20]: The `Short`, `Long` and `Examples` here should be the variables in functions/go/inflate-helm-chart/generated/docs.go

File: functions/go/inflate-helm-chart/main.go [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+		bytes, err := kyaml.Marshal(dataMap)
+		if err != nil {
+			return err
+		}
+		err = f.ConfigMapArgs(bytes)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("`functionConfig` must be 'ConfigMap' or 'InflateHelmChart'")
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/main.go:77]: Please use `backquote` for ConfigMap and InflateHelmChart

File: functions/go/inflate-helm-chart/main.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+		return nil, err
+	}
+	var rm resmap.ResMap
+	for _, p := range f.plugins {
+		rm, err = p.Generate()
+		if err != nil {
+			return nil, fmt.Errorf("failed to run generator: %w", err)
+		}
+
+		// check for duplicates for idempotency
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/main.go:95]: What's this about? Do we really need this?

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go [hunk 26]
```diff
@@ -77,31 +61,27 @@ func (p *HelmChartInflationGeneratorPlugin) establishTmpDir() (err error) {
...
 
-	// ChartHome might be consulted by the plugin (to read
-	// values files below it), so it must be located under
-	// the loader root (unless root restrictions are
-	// disabled, in which case this can be an absolute path).
+	// ChartHome might be written to by the function in a container,
+	// so it must be under the `/tmp` directory
 	if p.ChartHome == "" {
-		p.ChartHome = "charts"
+		p.ChartHome = "tmp/charts"
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go:72]: Should it be `/tmp/charts`?

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go [hunk 27]
```diff
@@ -73,50 +71,3 @@ type HelmChart struct {
 	// Defaults to 'false'.
 	IncludeCRDs bool `json:"includeCRDs,omitempty" yaml:"includeCRDs,omitempty"`
 }
-
-// HelmChartArgs contains arguments to helm.
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go:77]: I'd suggest leave these unchanged. Otherwise, it may be a pain when you try to rebase our fork to newer `kustomize/api`

**Comment by mengqiy** [comment]
I'd suggest add more e2e tests or examples to provide more coverage and exercise more supported options.
You can either do it in this PR or in a followup PR.

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go (hunk 26 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go:72]: It needs to be relative to run well as an exec function. If a user runs it with kustomize as an exec, it will use the local `./tmp/charts/` directory.

When run in a container, the `absPath` method is able to convert this to `/tmp/charts`.

File: functions/go/inflate-helm-chart/main.go (hunk 25 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/main.go:95]: The function won't be idempotent without it. Is that fine?

File: functions/go/inflate-helm-chart/go.mod (hunk 16 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/go.mod]: changed back to 3.12; no particular reason other than thats what was there before

File: functions/go/inflate-helm-chart/go.mod (hunk 17 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/go.mod]: changed it to just 0.8.11 (the latest release)

File: functions/go/inflate-helm-chart/go.sum (hunk 18 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/go.sum]: done

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go (hunk 19 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/builtins/HelmChartInflationGenerator.go:180]: no, it can be a relative path, e.g. ./tmp/my-values.yaml. I added examples to the table in the README per your comment below

File: functions/go/inflate-helm-chart/README.md (hunk 20 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: added examples! PTAL to see if it clarifies

File: functions/go/inflate-helm-chart/README.md [hunk 28]
```diff
@@ -0,0 +1,237 @@
...
+  repo: string
+  releaseName: string
+  namespace: string
+  valuesFile: string
+  includeCRDs: string
+```
+
+| Field        |  Description | 
+| -----------: |  ----------- | 
+`helmGlobals`  | Parameters applied to all Helm charts
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: removed

File: functions/go/inflate-helm-chart/README.md [hunk 29]
```diff
@@ -0,0 +1,237 @@
...
+`helmCharts`   | An array of helm chart parameters
+`chartHome`    | A filepath to a directory of charts. The function will look for the chart in this local directory before attempting to pull the chart from a specified repo. Defaults to "tmp/charts". When run in a container, this path MUST have the prefix "tmp/".
+`configHome`   | (exec only) Defines a value that the function should pass to helm via the HELM_CONFIG_HOME environment variable. If omitted, {tmpDir}/helm is used, where {tmpDir} is some temporary directory created by the function for the benefit of helm.
+`name`         | The name of the chart
+`version`      | The version of the chart
+`repo`         | A URL locating the chart on the internet
+`releaseName`  | Replaces RELEASE_NAME in the chart template output
+`namespace`    | Sets the target namespace for a release (.Release.Namespace in the template)
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: done

File: functions/go/inflate-helm-chart/README.md [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+`valuesFile`   | valuesFile is a remote or local file path to a values file to use _instead of_ the default values that accompanied the chart. The default values are in '{ChartHome}/{Name}/values.yaml'.
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: added a sentence to clarify

File: functions/go/inflate-helm-chart/README.md (hunk 21 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:199]: Added kustomize v4 to specify

File: functions/go/inflate-helm-chart/main.go (hunk 22 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/main.go:15]: done

File: functions/go/inflate-helm-chart/main.go (hunk 23 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/main.go]: fixed

File: functions/go/inflate-helm-chart/main.go (hunk 24 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/main.go]: fixed

File: functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go (hunk 27 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go]: thanks for the suggestion, I readded it

File: functions/go/inflate-helm-chart/README.md [hunk 31]
```diff
@@ -0,0 +1,238 @@
...
+`configHome`   | (exec only) Defines a value that the function should pass to helm via the HELM_CONFIG_HOME environment variable. If omitted, {tmpDir}/helm is used, where {tmpDir} is some temporary directory created by the function for the benefit of helm. | /tmp/helm/config
+`name`         | The name of the chart | minecraft
+`version`      | The version of the chart | 3.1.3
+`repo`         | A URL locating the chart on the internet | https://itzg.github.io/minecraft-server-charts
+`releaseName`  | Replaces RELEASE_NAME in the chart template output | test
+`namespace`    | Sets the target namespace for a release (`.Release.Namespace` in the template) | my-namespace
+`valuesFile`   | valuesFile is a remote or local file path to a values file to use instead of the default values that accompanied the chart. The default values are in '{chartHome}/{name}/values.yaml', where `chartHome` and `name` are the parameters defined above. | tmp/values.yaml <br> <br> https://raw.githubusercontent.com/config-sync-examples/helm-components/main/cert-manager-values.yaml
+`includeCRDs`  | Specifies if Helm should also generate CustomResourceDefinitions, defaults to 'false' | true
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:83]: > true

Please quote it, since it's a string.

File: functions/go/inflate-helm-chart/README.md [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+  releaseName: string
+  namespace: string
+  valuesFile: string
+  includeCRDs: string
+```
+
+
+| Field        |  Description | Example
+| -----------: |  ----------- | -----------
+`chartHome`    | A filepath to a directory of charts. The function will look for the chart in this local directory before attempting to pull the chart from a specified repo. Defaults to "tmp/charts". When run in a container, this path MUST have the prefix "tmp/". | tmp/charts
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:75]: Should it be `/tmp/charts`?
Please audit the doc ensure consistency?

File: functions/go/inflate-helm-chart/README.md [hunk 33]
```diff
@@ -0,0 +1,238 @@
...
+`configHome`   | (exec only) Defines a value that the function should pass to helm via the HELM_CONFIG_HOME environment variable. If omitted, {tmpDir}/helm is used, where {tmpDir} is some temporary directory created by the function for the benefit of helm. | /tmp/helm/config
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:76]: > (exec only)

This might be confusing for users that come from kpt side.
I'd suggest to say something like:
`This option is not supported when running in a container. It is supported only in exec mode (e.g. with kustomize).`

File: functions/go/inflate-helm-chart/README.md [hunk 34]
```diff
@@ -0,0 +1,238 @@
...
+  repo: string
+  releaseName: string
+  namespace: string
+  valuesFile: string
+  includeCRDs: string
+```
+
+
+| Field        |  Description | Example
+| -----------: |  ----------- | -----------
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:74]: I'm curious is the `:` doing something special here?

File: functions/go/inflate-helm-chart/README.md [hunk 35]
```diff
@@ -0,0 +1,238 @@
...
+`configHome`   | (exec only) Defines a value that the function should pass to helm via the HELM_CONFIG_HOME environment variable. If omitted, {tmpDir}/helm is used, where {tmpDir} is some temporary directory created by the function for the benefit of helm. | /tmp/helm/config
+`name`         | The name of the chart | minecraft
+`version`      | The version of the chart | 3.1.3
+`repo`         | A URL locating the chart on the internet | https://itzg.github.io/minecraft-server-charts
+`releaseName`  | Replaces RELEASE_NAME in the chart template output | test
+`namespace`    | Sets the target namespace for a release (`.Release.Namespace` in the template) | my-namespace
+`valuesFile`   | valuesFile is a remote or local file path to a values file to use instead of the default values that accompanied the chart. The default values are in '{chartHome}/{name}/values.yaml', where `chartHome` and `name` are the parameters defined above. | tmp/values.yaml <br> <br> https://raw.githubusercontent.com/config-sync-examples/helm-components/main/cert-manager-values.yaml
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:82]: > tmp/values.yaml <br> <br> https://raw.githubusercontent.com/config-sync-examples/helm-components/main/cert-manager-values.yaml

Are you trying to provide examples for local values and remote values? If so, I'd suggest state it clearly. e.g. 
```
Using a local values file: path/to/your/values.yaml <br> <br> Using a remote values file: https://raw.githubusercontent.com/config-sync-examples/helm-components/main/cert-manager-values.yaml
```

File: functions/go/inflate-helm-chart/README.md [hunk 36]
```diff
@@ -0,0 +1,238 @@
...
+  releaseName: string
+  namespace: string
+  valuesInline: map[string]interface{}
+  valuesFile: string
+  valuesMerge: string
+  includeCRDs: bool
+```
+
+| Field        |  Description | Example
+| -----------: |  ----------- | -----------
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:108]: Some comments above also apply here. Please audit thie table.

File: functions/go/inflate-helm-chart/README.md [hunk 37]
```diff
@@ -0,0 +1,238 @@
...
+          image: gcr.io/kpt-fn/inflate-helm-chart:unstable
+  data:
+    name: minecraft
+    repo: https://itzg.github.io/minecraft-server-charts
+    version: 3.1.3
+    releaseName: test
+```
+
+For both of the above kustomizations, you can use kustomize v4 to inflate
+the helm charts with the following command:
```
Comment (mengqiy) [inline functions/go/inflate-helm-chart/README.md:197]: Can you also link to the kustomize reference doc if any?

**Comment by mengqiy** [comment]
We should add deprecation message to the old TS function when we are done with golang one.

File: functions/go/inflate-helm-chart/README.md [hunk 38]
```diff
@@ -0,0 +1,238 @@
...
+or kustomize to inflate a specified helm chart.
+
+In kpt, the function can only be run imperatively. The function either
+needs network access to inflate a remote chart or needs a local file to be mounted
+into the container to inflate a local chart. As a result, to run the
+function with `kpt fn eval`, the flag `--network` must be used for remote charts,
+and the flag `--mount` must be used for local charts. See the examples for inflating
+[local] and [remote] charts.
+
+It can be used declaratively when run with kustomize. To run the function with kustomize,
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:43]: It does, with the command `kustomize fn run`. However this command is alpha, not documented anywhere, and there has been some discussion around removing it, so I am not sure if we should provide documentation or examples for it here.

File: functions/go/inflate-helm-chart/README.md (hunk 32 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:75]: Same as the above comment, I intentionally left it as `tmp/charts` because when using this as an exec function and inflating a local chart, this will allow the users to have the charts in their current directory by default. In a container environment, the `absPath()` method converts this to `/tmp/charts`.

If we want to use `/tmp/charts` by default instead, we can, but this will require kustomize users who use this as an exec function to specify an extra field for the path to their local charts.

File: functions/go/inflate-helm-chart/README.md (hunk 34 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:74]: It causes this column to be right-justified.

https://www.markdownguide.org/extended-syntax/#Alignment

File: functions/go/inflate-helm-chart/README.md (hunk 37 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:197]: There is no documentation for this in kustomize

File: functions/go/inflate-helm-chart/README.md (hunk 33 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: done

File: functions/go/inflate-helm-chart/README.md (hunk 35 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: done

File: functions/go/inflate-helm-chart/README.md (hunk 36 above)
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md:107]: I changed both tables according to your comments

**Comment by natasha41575** [comment]
> We should add deprecation message to the old TS function when we are done with golang one.

I can do this in my followup PR.

File: functions/go/inflate-helm-chart/README.md (hunk 34 above)
Comment (mengqiy) [reply functions/go/inflate-helm-chart/README.md:73]: Got it. Thanks

File: functions/go/inflate-helm-chart/README.md (hunk 37 above)
Comment (mengqiy) [reply functions/go/inflate-helm-chart/README.md:196]: Ack

File: functions/go/inflate-helm-chart/README.md (hunk 32 above)
Comment (mengqiy) [reply functions/go/inflate-helm-chart/README.md:74]: Sorry, I missed your earlier comment. There are a lot of comments here :P 

It seems your Dockerfile is using the default WORKDIR which is `/`. So it should be fine here. When in container, it resolves to `/tmp/charts`.

**Review by mengqiy — APPROVED** [review]

**Comment by mengqiy** [comment]
Please update the title.

**Comment by mengqiy** [comment]
Other stuff (more test (examples) coverage) can be done in followup PRs.
